File: LeetCode/Hard/2040-Kth_Smallest_Product_Of_Two_Sorted_Arrays/2040-KthSmallestProductOfTwoSortedArrays.cpp
class Solution {
public:
    std::pair<std::vector<int>, std::vector<int>> split(std::vector<int>& num) {
        std::vector<int> pos;
        std::vector<int> neg;

        for (int i = 0; i < num.size(); i++) {
            if (num[i] < 0) {
                neg.push_back(-num[i]);
            } else {
                pos.push_back(num[i]);
            }
        }

        std::reverse(neg.begin(), neg.end());

        return std::make_pair(pos, neg);
    }

    long long count_products(std::vector<int>& A, std::vector<int>& B, long long m) {
        long long count = 0;
        long long j = B.size() -1;

        for (int num :  A) {
            while (j >= 0 && (long long) num * B[j] > m) {
                j--;
            }

            count += j + 1;
        }

        return count;
    }

    long long kthSmallestProduct(vector<int>& nums1, vector<int>& nums2, long long k) {
        std::vector<int> A1, A2, B1, B2;

        std::pair< std::vector<int>,  std::vector<int>> result = split(nums1);
        A1 = result.first;
        A2 = result.second;

        result = split(nums2);
        B1 = result.first;
        B2 = result.second;

        int negative_count = A1.size() * B2.size() + A2.size() * B1.size();
        int  sign = 1;

        if (k > negative_count) {
            k -= negative_count;
        }
        else {
            k = negative_count - k + 1;
            sign = -1;
            std::swap(B1, B2);
        }

        long long left = 0;
        long long right = static_cast<long long>(pow(10, 10));

        while (left < right) {
            long long mid = (left + right) / 2;

            if (count_products(A1, B1, mid) + count_products(A2, B2, mid) >= k) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return sign * left;
    }
};

File: LeetCode/Hard/2040-Kth_Smallest_Product_Of_Two_Sorted_Arrays/2040-KthSmallestProductOfTwoSortedArrays.js
var split = function (nums) {
    let pos = [];
    let neg = [];

    for (let num of nums) {
        if (num < 0)
            neg.push(-num);
        else
            pos.push(num);
    }

    neg.reverse();
    return [pos, neg];
}

var count_products = function (A, B, m) {
    let count = 0n;
    let j = BigInt(B.length) - 1n;

    for (let num of A) {
        while (j >= 0n && BigInt(num) * BigInt(B[j]) > m) {
            j--;
        }

        count += j + 1n;
    }

    return count;
}

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number}
 */
var kthSmallestProduct = function (nums1, nums2, k) {
    let A1, A2, B1, B2;

    pair = split(nums1);
    A1 = pair[0];
    A2 = pair[1];

    pair = split(nums2);
    B1 = pair[0];
    B2 = pair[1];

    let count_negative = A1.length * B2.length + A2.length * B1.length;
    let sign = 1;

    if (k > count_negative) {
        k -= count_negative;
    } else {
        k = count_negative - k + 1;
        sign = -1;
        [B1, B2] = [B2, B1];
    }

    let left = 0n;
    let right = 10000000000n;

    while (left < right) {
        let mid = (left + right) / 2n;

        if (count_products(A1, B1, mid) + count_products(A2, B2, mid) >= BigInt(k)) {
            right = mid;
        } else {
            left = mid + 1n;
        }
    }

    return sign * Number(left);
};